core/src/avm1/object.rs

```diff
@@ -397,59 +336,12 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
     /// Set the interface list for this object. (Only useful for prototypes.)
     fn set_interfaces(&self, gc_context: MutationContext<'gc, '_>, iface_list: Vec<Object<'gc>>);
 
-    /// Determine if this object is an instance of a class.
-    ///
-    /// The class is provided in the form of its constructor function and the
-    /// explicit prototype of that constructor function. It is assumed that
-    /// they are already linked.
-    ///
-    /// Because ActionScript 2.0 added interfaces, this function cannot simply
-    /// check the prototype chain and call it a day. Each interface represents
-    /// a new, parallel prototype chain which also needs to be checked. You
-    /// can't implement interfaces within interfaces (fortunately), but if you
-    /// somehow could this would support that, too.
-    fn is_instance_of(
-        &self,
-        activation: &mut Activation<'_, 'gc, '_>,
-        constructor: Object<'gc>,
-        prototype: Object<'gc>,
-    ) -> Result<bool, Error<'gc>> {
-        let mut proto_stack = vec![];
-        if let Value::Object(p) = self.proto() {
-            proto_stack.push(p);
-        }
-
-        while let Some(this_proto) = proto_stack.pop() {
-            if Object::ptr_eq(this_proto, prototype) {
-                return Ok(true);
-            }
-
-            if let Value::Object(p) = this_proto.proto() {
-                proto_stack.push(p);
-            }
-
-            if activation.swf_version() >= 7 {
-                for interface in this_proto.interfaces() {
-                    if Object::ptr_eq(interface, constructor) {
-                        return Ok(true);
-                    }
-
-                    if let Value::Object(o) = interface.get("prototype", activation)? {
-                        proto_stack.push(o);
-                    }
-                }
-            }
-        }
-
-        Ok(false)
-    }
-
     /// Get the underlying script object, if it exists.
     fn as_script_object(&self) -> Option<ScriptObject<'gc>>;
 
-    /// Get the underlying array object, if it exists.
-    fn as_array_object(&self) -> Option<ArrayObject<'gc>> {
-        None
+    /// Check whether an underlying array object exists for this object.
+    fn is_array_object(&self) -> bool {
```

Comment:
Don't we want to keep `fn as_X(&self) -> Option<X>`, if only for consistency with other enum-downcast-functions?

Reply:
That was also my opinion when I introduced this function in #4611. I still think consistency is important, but on the other hand:
1. `is_array_object` is much more simple.
2. There's nothing in `ArrayObject` that cannot be retrieved using `as_script_object`.
3. In the future I plan to convert all of those `as_*` methods to a single generic one.

Reply:
Reverted for now.